build-logic/documentation/src/main/groovy/gradlebuild/docs/FindBadMultiLangSnippets.java

```diff
@@ -0,0 +1,215 @@
+ * Only rule enforced:
+ *   In a ==== ... ==== block, if there are at least two [.multi-language-sample] snippets,
+ *   flag when the first is GROOVY and the second is KOTLIN.
+ */
+@CacheableTask
+public abstract class FindBadMultiLangSnippets extends DefaultTask {
+
+    @InputDirectory
+    @PathSensitive(PathSensitivity.RELATIVE)
+    public abstract DirectoryProperty getDocumentationRoot();
+
+    @TaskAction
+    public void checkMultiLanguageSnippets() {
+        Map<File, List<Error>> errors = new TreeMap<>();
+
+        getDocumentationRoot().getAsFileTree().matching(spec -> spec.include("**/*.adoc")).forEach(file -> {
+            gatherBadSnippetsInFile(file, errors);
+        });
+
+        if (!errors.isEmpty()) {
+            StringBuilder sb = new StringBuilder("Found reversed multi-language snippet order in AsciiDoc files:\n");
+            for (Map.Entry<File, List<Error>> e : errors.entrySet()) {
+                sb.append(" - ").append(e.getKey().getPath()).append("\n");
+                for (Error err : e.getValue()) {
+                    sb.append("    line ").append(err.line).append(": ").append(err.message).append("\n");
+                }
+            }
+            throw new GradleException(sb.toString());
+        }
+    }
+
+    private void gatherBadSnippetsInFile(File file, Map<File, List<Error>> errors) {
+        List<String> lines = new ArrayList<>();
+        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
+            String l;
+            while ((l = br.readLine()) != null) {
+                lines.add(l);
+            }
+        } catch (IOException ex) {
+            // If a file can't be read
+            addError(errors, file, new Error(1, "Failed to read file: " + ex.getMessage()));
+            return;
+        }
+
+        boolean inExample = false; // between ==== ... ====
+        List<Snippet> currentSnippets = new ArrayList<>();
+
+        for (int i = 0; i < lines.size(); i++) {
+            String raw = lines.get(i);
+            String line = raw.trim();
+
+            if (line.equals("====")) {
+                // Toggle example block
+                if (!inExample) {
+                    inExample = true;
+                    currentSnippets.clear();
+                } else {
+                    // Closing an example block: check for reversed order only
+                    flagIfReversed(file, errors, currentSnippets);
+                    inExample = false;
+                    currentSnippets.clear();
+                }
+                continue;
+            }
+
+            if (!inExample) {
+                continue;
+            }
+
+            if (line.equals("[.multi-language-sample]")) {
+                int headerLine = i + 1; // 1-based
+                // Look ahead a bit to find [source, ...]
+                String sourceLang = null;
+                for (int j = i + 1; j < Math.min(i + 15, lines.size()); j++) {
```

Comment:
💭  Sketchy magic constant `15`. Giving it a name may help to read the procedural parsing a bit easier